app/data_models/questionnaire_store.py

```diff
@@ -1,23 +1,111 @@
 from app.questionnaire.rules.utils import parse_iso_8601_datetime
 from app.utilities.json import json_dumps, json_loads
-from app.utilities.types import SupplementaryDataListMapping
+from app.utilities.types import LocationType, SupplementaryDataListMapping
 
 if TYPE_CHECKING:
     from app.storage.encrypted_questionnaire_storage import (  # pragma: no cover
         EncryptedQuestionnaireStorage,
     )
 
 
+@dataclass
+class DataStores:
+    # self.metadata is a read-only view over self._metadata
+    metadata: MetadataProxy | None = None
+    response_metadata: MutableMapping = field(default_factory=dict)
+    list_store: ListStore = ListStore()
+    answer_store: AnswerStore = field(default_factory=AnswerStore)
+    progress_store: ProgressStore = ProgressStore()
+    supplementary_data_store: SupplementaryDataStore = SupplementaryDataStore()
+
+    def rule_evaluator(
+        self,
+        *,
+        schema: QuestionnaireSchema,
+        location: Location,
+        routing_path_block_ids: Iterable[str] | None = None,
+    ) -> rule_evaluator.RuleEvaluator:
+        return rule_evaluator.RuleEvaluator(
+            schema=schema,
+            answer_store=self.answer_store,
+            list_store=self.list_store,
+            metadata=self.metadata,
+            response_metadata=self.response_metadata,
+            progress_store=self.progress_store,
+            location=location,
+            routing_path_block_ids=routing_path_block_ids,
+            supplementary_data_store=self.supplementary_data_store,
+        )
+
+    def value_source_resolver(
+        self,
+        *,
+        schema: QuestionnaireSchema,
+        location: Location | RelationshipLocation | None,
+        list_item_id: str | None,
+        routing_path_block_ids: OrderedSet[str] | None = None,
+        assess_routing_path: bool = True,
+        use_default_answer: bool = True,
+        escape_answer_values: bool = True,
+    ) -> value_source_resolver.ValueSourceResolver:
+        return value_source_resolver.ValueSourceResolver(
+            answer_store=self.answer_store,
+            list_store=self.list_store,
+            metadata=self.metadata,
+            schema=schema,
+            location=location,
+            list_item_id=list_item_id,
+            escape_answer_values=escape_answer_values,
+            response_metadata=self.response_metadata,
+            use_default_answer=use_default_answer,
+            assess_routing_path=assess_routing_path,
+            routing_path_block_ids=routing_path_block_ids,
+            progress_store=self.progress_store,
+            supplementary_data_store=self.supplementary_data_store,
+        )
+
+    def placeholder_renderer(
+        self,
+        language: str,
+        schema: QuestionnaireSchema,
+        location: LocationType | None = None,
+        placeholder_preview_mode: bool | None = False,
+    ) -> placeholder_renderer.PlaceholderRenderer:
```

Comment:
it looks a bit odd to have the import name match the function name, with this class in a separate file I'd import `PlaceholderRenderer` directly

Reply:
Caused by circular import errors.